docs/proposals/pod-safety.md

```diff
@@ -0,0 +1,431 @@
+as **fencing** and is a well understood domain.
+
+This proposal covers the changes necessary to ensure:
+
+* Pet Sets can ensure **at most one** semantics for each individual pet
+* Other system components such as the node and namespace controller can
+  safely perform their responsibilities without violating that guarantee
+* An administrator or higher level controller can signal that a node
+  partition is permanent, allowing the Pet Set controller to proceed.
+* A fencing controller can take corrective action automatically to heal
+  partitions
+
+We will accomplish this by:
+
+* Clarifying which components are allowed to force delete pods (as opposed
+  to merely requesting termination)
+* Ensuring system components can observe partitioned pods and nodes
+  correctly
+* Defining how a fencing controller could safely interoperate with
+  partitioned nodes and pods to safely heal partitions
+* Describing how shared storage components without innate safety
+  guarantees can be safely shared on the cluster.
+
+
+### Current Guarantees for Pod lifecycle
+
+The existing pod model provides the following guarantees:
+
+* A pod is executed on exactly one node
+* A pod has the following lifecycle phases:
+  * Creation
+  * Scheduling
+  * Execution
+    * Init containers
+    * Application containers
+  * Termination
+  * Deletion
+* A pod can only move through its phases in order, and may not return
+  to an earlier phase.
+* A user may specify an interval on the pod called the **termination
+  grace period** that defines the minimum amount of time the pod will
+  have to complete the termination phase, and all components will honor
+  this interval.
+* Once a pod begins termination, its termination grace period can only
+  be shortened, not lengthened.
+
+Pod termination is divided into the following steps:
+
+* A component requests the termination of the pod by issuing a DELETE
+  to the pod resource with an optional **grace period**
+  * If no grace period is provided, the default from the pod is leveraged
+* When the kubelet observes the deletion, it starts a timer equal to the
+  grace period and performs the following actions:
+  * Executes the pre-stop hook, if specified, waiting up to **grace period**
+    seconds before continuing
+  * Sends the termination signal to the container runtime (SIGTERM or the
+    container image's STOPSIGNAL on Docker)
+  * Waits 2 seconds, or the remaining grace period, whichever is longer
+  * Sends the force termination signal to the container runtime (SIGKILL)
+* Once the kubelet observes the container is fully terminated, it issues
+  a status update to the REST API for the pod indicating termination, then
+  issues a DELETE with grace period = 0.
+
+If the kubelet crashes during the termination process, it will restart the
+termination process from the beginning (grace period is reset). This ensures
+that a process is always given **at least** grace period to terminate cleanly.
```

Comment:
Not sure what you mean by blessed and non-blessed part (terminology mismatch).  The API server is providing serializable guarantees, so there is only progress or not-progress w.r.t. to recorded changes.  The node controller will not do anything if split (can't make changes), but even then it simply requests termination.  The kubelet will observe the request and try to reconcile it, but not it will not handle behavior.

The fencer is allowed to force delete if it can:

1. Reach the master (not partitioned)
2. Guarantee total isolation of the process from the rest of the cluster (storage, network, quantum interference with other processes, whatever)

We definitely will have to consider sets of fencing (storage, network, and then power management), so 2 above is more nuanced than just "force delete" - it might be updating the PV record (as described elsewhere in this doc) as well as force deleting the pod.

If the fencer is wrong, or unsure, the fencer can't force delete the pod.